proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+
+```kotlin
+fetchById<User>("1182938") // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<User>("1182938", UserRepository()) // you can provide it manually.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. In this scenario, the following classes would be equivalent:
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class GroupRepository<A>(with val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> =
+    Group(R.loadAll())
+}
+
+// Regular Kotlin
+class GroupRepository<A>(val R: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Group<A> = with (R) {
+    Group(loadAll())
+  }
+}
+```
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): List<Group<A>> {
+    return listOf(Group(repoA.loadAll()))
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return Group(repoA.loadById(id))
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. Call site would be like:
+
+```kotlin
+fun fetchGroup<A>(with repo: GroupRepository<A>) = repo.loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
+}
+```
+
+We believe the encoding proposed above fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support compile time dependency resolution such as Scala where this is done via implicits.
+
+## Language changes
+
+* Add `with` to require evidence of extensions in both function and class/object declarations.
+* Add `extension` to provide instance evidence for a given interface.
+
+Usage of these language changes are demonstrated by the previous and below examples:
+
+#### Class constraint
+
+```kotlin
+extension class GroupRepository<A>(with R: Repository<A>) : Repository<Group<A>> {
+  /* ... */
+}
+```
+
+#### Function constraint
+
+```kotlin
+fun <A> fetch(id: String, with R: Repository<A>): A = loadById(id) // function position using parameter "R"
+```
+
+#### Extension evidence using an Object
```

Comment:
Do extensions on `fun` and `val` work? We need to add those cases here as well.

Reply:
Not yet, I'd say. This is part of what we discussed in the call yesterday, we're just supporting named extensions through explicit objects or classes for now. Can you confirm that @truizlop? If we require that then I'll add it to the "What needs to be done?" section.

Reply:
Exactly, right now we only support `extension class` and `extension object`, so if those cases need to be supported, they need to be added to the pending list. Is there any use case they cover, that is not covered with the currently supported options?

Reply:
I believe both `fun` and `val` should be supported as valid providers of extensions.
`fun` unlike `class` allows you to provide instances without incurring in allocations.
`val` unlike object can be parametric if it's a property and it's perhaps let's interesting than `fun`.

Reply:
How would those look like? I imagine they could support nested provisioning too, but what do you mean by not incur in allocations? You'd need to allocate for providing an instance, no matter what. You probably mean allocations prior to be called?

Reply:
I've added it to the TODO list in 297e65f. Can you review it @raulraja? Thx.